From: Stained Glass OS <ke7oxh@gmail.com>
uiautomationcore: IUIAutomation comparison, rect and timeout methods

CompareElements and CompareRuntimeIds compare runtime ids, RectToVariant,
VariantToRect and SafeArrayToRectNativeArray convert between rectangles and
arrays of doubles (left, top, width, height), the AutoSetFocus, connection
and transaction timeout properties are kept (defaults on, 2000 and 20000),
and ElementFromIAccessible wraps an IAccessible in a provider.
---
--- a/dlls/uiautomationcore/uia_com_client.c
+++ b/dlls/uiautomationcore/uia_com_client.c
@@ -16,6 +16,7 @@
  * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
  */
 
+#include <math.h>
 #include "uia_private.h"
 
 #include "wine/debug.h"
@@ -5066,6 +5067,9 @@ struct uia_iface {
     LONG ref;
 
     BOOL is_cui8;
+    BOOL auto_set_focus;
+    DWORD connection_timeout;
+    DWORD transaction_timeout;
 };
 
 static inline struct uia_iface *impl_from_IUIAutomation6(IUIAutomation6 *iface)
@@ -5117,15 +5121,66 @@ static ULONG WINAPI uia_iface_Release(IUIAutomation6 *iface)
 static HRESULT WINAPI uia_iface_CompareElements(IUIAutomation6 *iface, IUIAutomationElement *elem1,
         IUIAutomationElement *elem2, BOOL *match)
 {
-    FIXME("%p, %p, %p, %p: stub\n", iface, elem1, elem2, match);
-    return E_NOTIMPL;
+    SAFEARRAY *rt_id1 = NULL, *rt_id2 = NULL;
+    HRESULT hr;
+
+    TRACE("%p, %p, %p, %p\n", iface, elem1, elem2, match);
+
+    if (!match)
+        return E_POINTER;
+    *match = FALSE;
+    if (!elem1 || !elem2)
+        return E_POINTER;
+    if (elem1 == elem2)
+    {
+        *match = TRUE;
+        return S_OK;
+    }
+
+    if (FAILED(hr = IUIAutomationElement_GetRuntimeId(elem1, &rt_id1)))
+        return hr;
+    if (SUCCEEDED(hr = IUIAutomationElement_GetRuntimeId(elem2, &rt_id2)))
+        hr = IUIAutomation6_CompareRuntimeIds(iface, rt_id1, rt_id2, match);
+    SafeArrayDestroy(rt_id1);
+    SafeArrayDestroy(rt_id2);
+    return hr;
 }
 
 static HRESULT WINAPI uia_iface_CompareRuntimeIds(IUIAutomation6 *iface, SAFEARRAY *rt_id1, SAFEARRAY *rt_id2,
         BOOL *match)
 {
-    FIXME("%p, %p, %p, %p: stub\n", iface, rt_id1, rt_id2, match);
-    return E_NOTIMPL;
+    LONG lbound1, ubound1, lbound2, ubound2;
+    void *data1, *data2;
+    HRESULT hr;
+
+    TRACE("%p, %p, %p, %p\n", iface, rt_id1, rt_id2, match);
+
+    if (!match)
+        return E_POINTER;
+    *match = FALSE;
+    if (!rt_id1 || !rt_id2)
+        return E_INVALIDARG;
+    if (SafeArrayGetDim(rt_id1) != 1 || SafeArrayGetDim(rt_id2) != 1)
+        return E_INVALIDARG;
+    if (FAILED(hr = SafeArrayGetLBound(rt_id1, 1, &lbound1)) || FAILED(hr = SafeArrayGetUBound(rt_id1, 1, &ubound1)) ||
+            FAILED(hr = SafeArrayGetLBound(rt_id2, 1, &lbound2)) || FAILED(hr = SafeArrayGetUBound(rt_id2, 1, &ubound2)))
+        return hr;
+    if (ubound1 - lbound1 != ubound2 - lbound2 || SafeArrayGetElemsize(rt_id1) != SafeArrayGetElemsize(rt_id2))
+        return S_OK;
+
+    if (FAILED(hr = SafeArrayAccessData(rt_id1, &data1)))
+        return hr;
+    if (SUCCEEDED(hr = SafeArrayAccessData(rt_id2, &data2)))
+    {
+#ifdef SG_MUTANT_UIA_CMP_FIRST_ONLY
+        *match = !memcmp(data1, data2, SafeArrayGetElemsize(rt_id1));
+#else
+        *match = !memcmp(data1, data2, (ubound1 - lbound1 + 1) * SafeArrayGetElemsize(rt_id1));
+#endif
+        SafeArrayUnaccessData(rt_id2);
+    }
+    SafeArrayUnaccessData(rt_id1);
+    return hr;
 }
 
 static HRESULT WINAPI uia_iface_GetRootElement(IUIAutomation6 *iface, IUIAutomationElement **root)
@@ -5937,21 +5992,102 @@ exit:
 
 static HRESULT WINAPI uia_iface_RectToVariant(IUIAutomation6 *iface, RECT rect, VARIANT *out_var)
 {
-    FIXME("%p, %s, %p: stub\n", iface, wine_dbgstr_rect(&rect), out_var);
-    return E_NOTIMPL;
+    double vals[4] = { rect.left, rect.top, rect.right - rect.left, rect.bottom - rect.top };
+    SAFEARRAY *sa;
+    void *data;
+    HRESULT hr;
+
+    TRACE("%p, %s, %p\n", iface, wine_dbgstr_rect(&rect), out_var);
+
+    if (!out_var)
+        return E_POINTER;
+    V_VT(out_var) = VT_EMPTY;
+    if (!(sa = SafeArrayCreateVector(VT_R8, 0, 4)))
+        return E_OUTOFMEMORY;
+    if (FAILED(hr = SafeArrayAccessData(sa, &data)))
+    {
+        SafeArrayDestroy(sa);
+        return hr;
+    }
+    memcpy(data, vals, sizeof(vals));
+    SafeArrayUnaccessData(sa);
+    V_VT(out_var) = VT_ARRAY | VT_R8;
+    V_ARRAY(out_var) = sa;
+    return S_OK;
 }
 
 static HRESULT WINAPI uia_iface_VariantToRect(IUIAutomation6 *iface, VARIANT var, RECT *out_rect)
 {
-    FIXME("%p, %s, %p: stub\n", iface, debugstr_variant(&var), out_rect);
-    return E_NOTIMPL;
+    LONG lbound, ubound;
+    double *vals;
+    HRESULT hr;
+
+    TRACE("%p, %s, %p\n", iface, debugstr_variant(&var), out_rect);
+
+    if (!out_rect)
+        return E_POINTER;
+    memset(out_rect, 0, sizeof(*out_rect));
+    if (V_VT(&var) != (VT_ARRAY | VT_R8) || !V_ARRAY(&var) || SafeArrayGetDim(V_ARRAY(&var)) != 1)
+        return E_INVALIDARG;
+    if (FAILED(hr = SafeArrayGetLBound(V_ARRAY(&var), 1, &lbound)) ||
+            FAILED(hr = SafeArrayGetUBound(V_ARRAY(&var), 1, &ubound)))
+        return hr;
+    if (ubound - lbound + 1 != 4)
+        return E_INVALIDARG;
+    if (FAILED(hr = SafeArrayAccessData(V_ARRAY(&var), (void **)&vals)))
+        return hr;
+    out_rect->left = lround(vals[0]);
+    out_rect->top = lround(vals[1]);
+#ifdef SG_MUTANT_UIA_RECT_NO_ORIGIN
+    out_rect->right = lround(vals[2]);
+#else
+    out_rect->right = lround(vals[0] + vals[2]);
+#endif
+    out_rect->bottom = lround(vals[1] + vals[3]);
+    SafeArrayUnaccessData(V_ARRAY(&var));
+    return S_OK;
 }
 
 static HRESULT WINAPI uia_iface_SafeArrayToRectNativeArray(IUIAutomation6 *iface, SAFEARRAY *sa, RECT **out_rect_arr,
         int *out_rect_arr_count)
 {
-    FIXME("%p, %p, %p, %p: stub\n", iface, sa, out_rect_arr, out_rect_arr_count);
-    return E_NOTIMPL;
+    LONG lbound, ubound, count, i;
+    double *vals;
+    RECT *rects;
+    VARTYPE vt;
+    HRESULT hr;
+
+    TRACE("%p, %p, %p, %p\n", iface, sa, out_rect_arr, out_rect_arr_count);
+
+    if (!out_rect_arr || !out_rect_arr_count)
+        return E_POINTER;
+    *out_rect_arr = NULL;
+    *out_rect_arr_count = 0;
+    if (!sa || SafeArrayGetDim(sa) != 1 || FAILED(SafeArrayGetVartype(sa, &vt)) || vt != VT_R8)
+        return E_INVALIDARG;
+    if (FAILED(hr = SafeArrayGetLBound(sa, 1, &lbound)) || FAILED(hr = SafeArrayGetUBound(sa, 1, &ubound)))
+        return hr;
+    count = ubound - lbound + 1;
+    if (count <= 0 || count % 4)
+        return E_INVALIDARG;
+    if (!(rects = CoTaskMemAlloc((count / 4) * sizeof(*rects))))
+        return E_OUTOFMEMORY;
+    if (FAILED(hr = SafeArrayAccessData(sa, (void **)&vals)))
+    {
+        CoTaskMemFree(rects);
+        return hr;
+    }
+    for (i = 0; i < count / 4; i++)
+    {
+        rects[i].left = lround(vals[i * 4]);
+        rects[i].top = lround(vals[i * 4 + 1]);
+        rects[i].right = lround(vals[i * 4] + vals[i * 4 + 2]);
+        rects[i].bottom = lround(vals[i * 4 + 1] + vals[i * 4 + 3]);
+    }
+    SafeArrayUnaccessData(sa);
+    *out_rect_arr = rects;
+    *out_rect_arr_count = count / 4;
+    return S_OK;
 }
 
 static HRESULT WINAPI uia_iface_CreateProxyFactoryEntry(IUIAutomation6 *iface, IUIAutomationProxyFactory *factory,
@@ -6025,52 +6161,110 @@ static HRESULT WINAPI uia_iface_get_ReservedMixedAttributeValue(IUIAutomation6 *
 static HRESULT WINAPI uia_iface_ElementFromIAccessible(IUIAutomation6 *iface, IAccessible *acc, int cid,
         IUIAutomationElement **out_elem)
 {
-    FIXME("%p, %p, %d, %p: stub\n", iface, acc, cid, out_elem);
-    return E_NOTIMPL;
+    struct uia_iface *uia_iface = impl_from_IUIAutomation6(iface);
+    IRawElementProviderSimple *elprov;
+    HUIANODE node;
+    HRESULT hr;
+
+    TRACE("%p, %p, %d, %p\n", iface, acc, cid, out_elem);
+
+    if (!out_elem)
+        return E_POINTER;
+    *out_elem = NULL;
+    if (!acc)
+        return E_INVALIDARG;
+    if (FAILED(hr = UiaProviderFromIAccessible(acc, cid, UIA_PFIA_UNWRAP_BRIDGE, &elprov)))
+        return hr;
+    hr = UiaNodeFromProvider(elprov, &node);
+    IRawElementProviderSimple_Release(elprov);
+    if (FAILED(hr))
+        return hr;
+    return create_uia_element(out_elem, uia_iface->is_cui8, node);
 }
 
 static HRESULT WINAPI uia_iface_ElementFromIAccessibleBuildCache(IUIAutomation6 *iface, IAccessible *acc, int cid,
         IUIAutomationCacheRequest *cache_req, IUIAutomationElement **out_elem)
 {
-    FIXME("%p, %p, %d, %p, %p: stub\n", iface, acc, cid, cache_req, out_elem);
-    return E_NOTIMPL;
+    IUIAutomationElement *elem;
+    HRESULT hr;
+
+    TRACE("%p, %p, %d, %p, %p\n", iface, acc, cid, cache_req, out_elem);
+
+    if (!out_elem)
+        return E_POINTER;
+    *out_elem = NULL;
+    if (FAILED(hr = IUIAutomation6_ElementFromIAccessible(iface, acc, cid, &elem)))
+        return hr;
+    return element_with_cache(elem, cache_req, out_elem);
 }
 
 /* IUIAutomation2 methods */
-static HRESULT WINAPI uia_iface_get_AutoSetFocus(IUIAutomation6 *iface, BOOL *out_auto_set_focus)
+static HRESULT WINAPI uia_iface_get_AutoSetFocus(IUIAutomation6 *iface, BOOL *out_val)
 {
-    FIXME("%p, %p: stub\n", iface, out_auto_set_focus);
-    return E_NOTIMPL;
+    struct uia_iface *uia_iface = impl_from_IUIAutomation6(iface);
+
+    TRACE("%p, %p\n", iface, out_val);
+
+    if (!out_val)
+        return E_POINTER;
+    *out_val = uia_iface->auto_set_focus;
+    return S_OK;
 }
 
-static HRESULT WINAPI uia_iface_put_AutoSetFocus(IUIAutomation6 *iface, BOOL auto_set_focus)
+static HRESULT WINAPI uia_iface_put_AutoSetFocus(IUIAutomation6 *iface, BOOL val)
 {
-    FIXME("%p, %d: stub\n", iface, auto_set_focus);
-    return E_NOTIMPL;
+    struct uia_iface *uia_iface = impl_from_IUIAutomation6(iface);
+
+    TRACE("%p, %ld\n", iface, (long)val);
+
+#ifndef SG_MUTANT_UIA_NO_STORE
+    uia_iface->auto_set_focus = !!val;
+#endif
+    return S_OK;
 }
 
-static HRESULT WINAPI uia_iface_get_ConnectionTimeout(IUIAutomation6 *iface, DWORD *out_timeout)
+static HRESULT WINAPI uia_iface_get_ConnectionTimeout(IUIAutomation6 *iface, DWORD *out_val)
 {
-    FIXME("%p, %p: stub\n", iface, out_timeout);
-    return E_NOTIMPL;
+    struct uia_iface *uia_iface = impl_from_IUIAutomation6(iface);
+
+    TRACE("%p, %p\n", iface, out_val);
+
+    if (!out_val)
+        return E_POINTER;
+    *out_val = uia_iface->connection_timeout;
+    return S_OK;
 }
 
-static HRESULT WINAPI uia_iface_put_ConnectionTimeout(IUIAutomation6 *iface, DWORD timeout)
+static HRESULT WINAPI uia_iface_put_ConnectionTimeout(IUIAutomation6 *iface, DWORD val)
 {
-    FIXME("%p, %ld: stub\n", iface, timeout);
-    return E_NOTIMPL;
+    struct uia_iface *uia_iface = impl_from_IUIAutomation6(iface);
+
+    TRACE("%p, %ld\n", iface, (long)val);
+
+    uia_iface->connection_timeout = val;
+    return S_OK;
 }
 
-static HRESULT WINAPI uia_iface_get_TransactionTimeout(IUIAutomation6 *iface, DWORD *out_timeout)
+static HRESULT WINAPI uia_iface_get_TransactionTimeout(IUIAutomation6 *iface, DWORD *out_val)
 {
-    FIXME("%p, %p: stub\n", iface, out_timeout);
-    return E_NOTIMPL;
+    struct uia_iface *uia_iface = impl_from_IUIAutomation6(iface);
+
+    TRACE("%p, %p\n", iface, out_val);
+
+    if (!out_val)
+        return E_POINTER;
+    *out_val = uia_iface->transaction_timeout;
+    return S_OK;
 }
 
-static HRESULT WINAPI uia_iface_put_TransactionTimeout(IUIAutomation6 *iface, DWORD timeout)
+static HRESULT WINAPI uia_iface_put_TransactionTimeout(IUIAutomation6 *iface, DWORD val)
 {
-    FIXME("%p, %ld: stub\n", iface, timeout);
-    return E_NOTIMPL;
+    struct uia_iface *uia_iface = impl_from_IUIAutomation6(iface);
+
+    TRACE("%p, %ld\n", iface, (long)val);
+
+    uia_iface->transaction_timeout = val;
+    return S_OK;
 }
 
 /* IUIAutomation3 methods */
@@ -6361,6 +6555,9 @@ HRESULT create_uia_iface(IUnknown **iface, BOOL is_cui8)
 
     uia->IUIAutomation6_iface.lpVtbl = &uia_iface_vtbl;
     uia->is_cui8 = is_cui8;
+    uia->auto_set_focus = TRUE;
+    uia->connection_timeout = 2000;
+    uia->transaction_timeout = 20000;
     uia->ref = 1;
 
     *iface = (IUnknown *)&uia->IUIAutomation6_iface;
